sql/analyzer/symbol_resolution.go

```diff
@@ -0,0 +1,286 @@
+// needed aliased columns qualified with the base table name,
+// and stars if applicable.
+// TODO: we don't have any tests with the unqualified confition
+func gatherTableAlias(
+	n sql.Node,
+	needed map[tableCol]int,
+	stars map[string]struct{},
+	unqualifiedStar bool,
+) ([]tableCol, []string) {
+	var cols []tableCol
+	var nodeStars []string
+	switch n := n.(type) {
+	case *plan.TableAlias:
+		alias := n.Name()
+		var base string
+		if rt := seeThroughDecoration(n.Child); rt != nil {
+			base = rt.Name()
+		}
+		_, starred := stars[alias]
+		if unqualifiedStar {
+			starred = true
+		}
+		for _, col := range n.Schema() {
+			baseCol := tableCol{table: base, col: col.Name}
+			aliasCol := tableCol{table: alias, col: col.Name}
+			if starred || needed[aliasCol] > 0 {
+				// if the outer scope requests an aliased column
+				// a table lower in the tree must provide the source
+				cols = append(cols, baseCol)
+			}
+		}
+		for t := range stars {
+			if t == alias {
+				nodeStars = append(nodeStars, base)
+			}
+		}
+		return cols, nodeStars
+	default:
+	}
+	return cols, nodeStars
+}
+
+// todo(max): implement this
+func gatherSubqueryExpression(n sql.Node) ([]tableCol, []string, bool) {
+	if sq := findSubqueryExpr(n); sq != nil {
+		return gatherOuterCols(sq.Query)
+	}
+	return nil, nil, false
+}
+
+// pruneTables removes unneeded columns from *plan.ResolvedTable nodes
+func pruneTables(ctx *sql.Context, a *Analyzer, n sql.Node, s *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	needed := make(map[tableCol]int)
+	var unqualifiedStar bool
+	stars := make(map[string]struct{})
+
+	push := func(cols []tableCol, nodeStars []string, nodeUnq bool) {
+		for _, c := range cols {
+			needed[c]++
+		}
+		for _, c := range nodeStars {
+			stars[c] = struct{}{}
+		}
+		unqualifiedStar = unqualifiedStar || nodeUnq
+	}
+
+	pop := func(cols []tableCol, nodeStars []string, beforeUnq bool) {
+		for _, c := range cols {
+			needed[c]--
+		}
+		for _, c := range nodeStars {
+			delete(stars, c)
+		}
+		unqualifiedStar = beforeUnq
+	}
+
+	// preOrder walk constructs a new tree. Nodes pass dependencies
+	// to children, and reset dependencies before returning to parent.
+	//
+	// The dependencies considered are:
+	//  - outerCols: columns used by filters or other expressions
+	//    sourced from outside the node
+	//  - aliasCols: a bridge between outside columns and an aliased
+	//    data source.
+	//  - subqueryCols: correlated subqueries have outside cols not
+	//    satisfied by tablescans in the subquery
+	//
+	// Stars are handled similarly, but with broader scoping authority.
+	var pruneWalk func(n sql.Node) (sql.Node, transform.TreeIdentity, error)
+	pruneWalk = func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch n := n.(type) {
+		case *plan.ResolvedTable:
+			return pruneTableCols(n, needed, stars, unqualifiedStar)
+		case sql.OpaqueNode, *plan.InsertInto, *plan.DeleteFrom, *plan.Update,
```

Comment:
This is the kind of rule that's going to break over time

Rather than having such a huge blacklist, is there a way we can narrow the scope to operate on exactly N kinds of nodes and skip everything else?

Reply:
yeah that's a lot better